README.md
# RusPiRo cache maintenance crate

This crates provides several cach maintenance functions that helps clearing or invalidating the cache of the Raspberry Pi.
Especially when it comes to cross core and core to GPU communications (like mailbox calls) the cache need to be cleared/invalidated
to ensure access to the most recent values stores in the memory.

[![Travis-CI Status](https://api.travis-ci.com/RusPiRo/ruspiro-cache.svg?branch=release)](https://travis-ci.com/RusPiRo/ruspiro-cache)
[![Latest Version](https://img.shields.io/crates/v/ruspiro-cache.svg)](https://crates.io/crates/ruspiro-cache)
[![Documentation](https://docs.rs/ruspiro-cache/badge.svg)](https://docs.rs/ruspiro-cache)
[![License](https://img.shields.io/crates/l/ruspiro-cache.svg)](https://github.com/RusPiRo/ruspiro-cache#license)

## Usage

To use the crate just add the following dependency to your ``Cargo.toml`` file:

```toml
[dependencies]
ruspiro-cache = "||VERSION||"
```

Once done the access to the cache maintenance functions is available like so:

```rust
use ruspiro-cache as cache;

fn demo() {
    cache::clean(); // clean the data cache
    cache::invalidate(); // invalidate the data cache
    cache::flush(); // clean and invalidate the data cache
}
```

## License

Licensed under Apache License, Version 2.0, ([LICENSE-APACHE](LICENSE-APACHE) or http://www.apache.org/licenses/LICENSE-2.0) or MIT ([LICENSE-MIT](LICENSE-MIT) or http://opensource.org/licenses/MIT)) at your choice.


CHANGELOG.md
# Changelog

## :peach: v0.4.0

This version removes the aarch32 support from the crate and introduces the use of a better pipeline config to build and publish this crate.

- ### :wrench: Maintenance

  - refactor the crate and remove aarch32 support
  - remove the assembly code and use rust code to do the cache maintenance operations with a bit of inline assembly
  - introduce the new pipeline configuration

## :banana: v0.3.1
  
- ### :wrench: Maintenance

  - use `cargo make` to stabilize build
  - remove `asm` feature as the crate does not use inline assembly

src/lib.rs
/***********************************************************************************************************************
 * Copyright (c) 2020 by the authors
 *
 * Author: André Borrmann <pspwizard@gmx.de>
 * License: Apache License 2.0 / MIT
 **********************************************************************************************************************/
#![doc(html_root_url = "https://docs.rs/ruspiro-cache/||VERSION||")]
#![cfg_attr(not(any(test, doctest)), no_std)]
#![feature(llvm_asm, core_intrinsics)]
// this crate does only compile to something usefull if targeted on Aarch64
#![cfg(target_arch = "aarch64")]

//! # Cache Maintenance Operations
//!
//! If the caches are active on the Raspberry Pi than there might be specific cache operations needed to clean and
//! invalidate the cache to ensure in cross core and/or ARM core to GPU communications the most recent data is seen.
//!
//! # Usage
//!
//! ```no_run
//! use ruspiro_cache as cache;
//!
//! fn doc() {
//!     cache::clean(); // clean the data cache
//!     cache::invalidate(); // invalidate the data cache
//!     cache::cleaninvalidate(); // clean and invalidate the data cache
//! }
//! ```

mod dcache;
pub use dcache::*;

/// Perform a cache clean operation on the entire data cache
pub fn clean() {
    unsafe { dcache::clean_dcache() }
}

/// Perform a cache invalidate operation on the entire data cache
pub fn invalidate() {
    unsafe { dcache::invalidate_dcache() }
}

/// Perform a cache clean and invalidate operation on the entire data cache
pub fn flush() {
    unsafe { dcache::flush_dcache() }
}


src/dcache.rs
/***********************************************************************************************************************
 * Copyright (c) 2020 by the authors
 *
 * Author: André Borrmann <pspwizard@gmx.de>
 * License: Apache License 2.0 / MIT
 **********************************************************************************************************************/

//! # Data Cache Operations
//!

use ruspiro_arch_aarch64::register::{
    el0::ctr_el0, el1::ccsidr_el1, el1::clidr_el1, el1::csselr_el1,
};

use ruspiro_arch_aarch64::instructions::{dsb, isb};

#[derive(Clone)]
enum CacheOperation {
    Clean,
    Invalidate,
    Flush,
}

#[derive(Clone, Copy)]
#[repr(u64)]
enum CacheLevel {
    L1 = 0,
    L2 = 1,
    L3 = 2,
}

pub unsafe fn invalidate_dcache() {
    dsb();

    // get level of coherency
    let loc = clidr_el1::read(clidr_el1::LOC::Field);
    if loc.value() != 0 {
        maintain_dcache(CacheOperation::Invalidate, CacheLevel::L1);
        maintain_dcache(CacheOperation::Invalidate, CacheLevel::L2);
        maintain_dcache(CacheOperation::Invalidate, CacheLevel::L3);
    }

    csselr_el1::write(csselr_el1::LEVEL::L1);
    dsb();
    isb();
}

pub unsafe fn clean_dcache() {
    dsb();

    // get level of coherency
    let loc = clidr_el1::read(clidr_el1::LOC::Field);
    if loc.value() != 0 {
        maintain_dcache(CacheOperation::Clean, CacheLevel::L1);
        maintain_dcache(CacheOperation::Clean, CacheLevel::L2);
        maintain_dcache(CacheOperation::Clean, CacheLevel::L3);
    }

    csselr_el1::write(csselr_el1::LEVEL::L1);
    dsb();
    isb();
}

pub unsafe fn flush_dcache() {
    dsb();

    // get level of coherency
    let loc = clidr_el1::read(clidr_el1::LOC::Field);
    if loc.value() != 0 {
        maintain_dcache(CacheOperation::Flush, CacheLevel::L1);
        maintain_dcache(CacheOperation::Flush, CacheLevel::L2);
        maintain_dcache(CacheOperation::Flush, CacheLevel::L3);
    }

    csselr_el1::write(csselr_el1::LEVEL::L1);
    dsb();
    isb();
}

unsafe fn maintain_dcache(operation: CacheOperation, level: CacheLevel) {
    // get the cache type for the requested cache level
    let cache_type = match level {
        CacheLevel::L1 => clidr_el1::read(clidr_el1::CTYPE1::Field).value(),
        CacheLevel::L2 => clidr_el1::read(clidr_el1::CTYPE2::Field).value(),
        CacheLevel::L3 => clidr_el1::read(clidr_el1::CTYPE3::Field).value(),
    };
    // if cache type is "no cache" or "instruction cache only" nothing to do, otherwise
    // process this cache leval
    if cache_type >= 0x2 {
        // select the cache level for the cache operations
        match level {
            CacheLevel::L1 => csselr_el1::write(csselr_el1::LEVEL::L1),
            CacheLevel::L2 => csselr_el1::write(csselr_el1::LEVEL::L2),
            CacheLevel::L3 => csselr_el1::write(csselr_el1::LEVEL::L3),
        }
        // instruction barrier to ensure the cache level has been choosen be the
        // previous instruction
        isb();
        let cache_line_size = ccsidr_el1::read(ccsidr_el1::LINESIZE::Field).value();
        let cache_line_size_x2 = cache_line_size + 4;
        let assoc = ccsidr_el1::read(ccsidr_el1::ASSOC::Field).value();
        // TODO: find bit position clz w5, w4?
        let assoc_bit_x5 = core::intrinsics::ctlz(assoc); //??
        let num_sets_x7 = ccsidr_el1::read(ccsidr_el1::NUMSETS::Field).value();
        for num_sets in (0..num_sets_x7).rev() {
            for way in (0..assoc).rev() {
                let x6 = way << assoc_bit_x5;
                let x11 = ((level as u64) << 1) | x6;
                let x6 = num_sets << cache_line_size_x2;
                let x11 = x11 | x6;
                // invalidate data cache by set/way
                match operation {
                    CacheOperation::Invalidate => llvm_asm!("dc isw, $0"::"r"(x11)::"volatile"),
                    CacheOperation::Clean => llvm_asm!("dc csw, $0"::"r"(x11)::"volatile"),
                    CacheOperation::Flush => llvm_asm!("dc cisw, $0"::"r"(x11)::"volatile"),
                }
            }
        }
    }
}

pub unsafe fn flush_dcache_range(from: usize, size: usize) {
    dsb();
    let dcls = dcache_line_size();
    let end = from + size;
    let start = from & !(dcls - 1);
    let mut current = start;
    while current < end {
        // clean & invalidate D line / unified line
        llvm_asm!("dc civac, $0"::"r"(current)::"volatile");
        current += dcls;
    }
    dsb();
}

unsafe fn dcache_line_size() -> usize {
    //let mut cls: usize;
    let dcls_log2 = ctr_el0::read(ctr_el0::DminLine::Field);
    (4 << dcls_log2.value()) as usize
}


src/asm/cache.s
/***********************************************************************************************************************
 * Raspberry Pi3 cache operation functions
 * cache opeartions are needed as the MMU need an active MMU and several
 * cross core/cross GPU operations need to clean/invalite the cache to se the
 * most actual data in the memory
 * TODO: actuall always the whole cache gets cleaned/invalidated, for better performance
 *       this should be enabled for specific address regions.
 *
 * Copyright (c) 2019 by the authors
 *
 * Author: André Borrmann
 * License: Apache License 2.0
 **********************************************************************************************************************/

.global __invalidate_dcache
.global __clean_dcache
.global __cleaninvalidate_dcache

/*
 *************************************************************************
 *
 * invalidate_dcache - invalidate the entire d-cache by set/way
 *
 * Note: for Cortex-A53, there is no cp instruction for invalidating
 * the whole D-cache. Need to invalidate each line.
 *
 *************************************************************************
 */

__invalidate_dcache:
	push	{r0 - r12, lr}

	mrc	p15, 1, r0, c0, c0, 1		/* read CLIDR */
	ands	r3, r0, #0x7000000
	mov	r3, r3, lsr #23			/* cache level value (naturally aligned) */
	beq	.ifinished
	mov	r10, #0				/* start with level 0 */
.iloop1:
	add	r2, r10, r10, lsr #1		/* work out 3xcachelevel */
	mov	r1, r0, lsr r2			/* bottom 3 bits are the Cache type for this level */
	and	r1, r1, #7			/* get those 3 bits alone */
	cmp	r1, #2
	blt	.iskip				/* no cache or only instruction cache at this level */
	mcr	p15, 2, r10, c0, c0, 0		/* write the Cache Size selection register */
	isb					/* isb to sync the change to the CacheSizeID reg */
	mrc	p15, 1, r1, c0, c0, 0		/* reads current Cache Size ID register */
	and	r2, r1, #7			/* extract the line length field */
	add	r2, r2, #4			/* add 4 for the line length offset (log2 16 bytes) */
	ldr	r4, =0x3ff
	ands	r4, r4, r1, lsr #3		/* r4 is the max number on the way size (right aligned) */
	clz	r5, r4				/* r5 is the bit position of the way size increment */
	ldr	r7, =0x7fff
	ands	r7, r7, r1, lsr #13		/* r7 is the max number of the index size (right aligned) */
.iloop2:
	mov	r9, r4				/* r9 working copy of the max way size (right aligned) */
.iloop3:
	orr	r11, r10, r9, lsl r5		/* factor in the way number and cache number into r11 */
	orr	r11, r11, r7, lsl r2		/* factor in the index number */

	mcr	p15, 0, r11, c7, c6, 2		// invalidate by set/way

	subs	r9, r9, #1			/* decrement the way number */
	bge	.iloop3
	subs	r7, r7, #1			/* decrement the index */
	bge	.iloop2
.iskip:
	add	r10, r10, #2			/* increment the cache number */
	cmp	r3, r10
	bgt	.iloop1

.ifinished:
	mov	r10, #0				/* swith back to cache level 0 */
	mcr	p15, 2, r10, c0, c0, 0		/* select current cache level in cssr */
	dsb
	isb

	pop {r0-r12, lr}
	bx	lr


/*
 *************************************************************************
 *
 * clean_dcache - clean the entire d-cache by set/way
 *
 * Note: for Cortex-A53, there is no cp instruction for invalidating
 * the whole D-cache. Need to invalidate each line.
 *
 *************************************************************************
 */

__clean_dcache:
	push	{r0 - r12, lr}

	mrc	p15, 1, r0, c0, c0, 1		/* read CLIDR */
	ands	r3, r0, #0x7000000
	mov	r3, r3, lsr #23			/* cache level value (naturally aligned) */
	beq	.cfinished
	mov	r10, #0				/* start with level 0 */
.cloop1:
	add	r2, r10, r10, lsr #1		/* work out 3xcachelevel */
	mov	r1, r0, lsr r2			/* bottom 3 bits are the Cache type for this level */
	and	r1, r1, #7			/* get those 3 bits alone */
	cmp	r1, #2
	blt	.cskip				/* no cache or only instruction cache at this level */
	mcr	p15, 2, r10, c0, c0, 0		/* write the Cache Size selection register */
	isb					/* isb to sync the change to the CacheSizeID reg */
	mrc	p15, 1, r1, c0, c0, 0		/* reads current Cache Size ID register */
	and	r2, r1, #7			/* extract the line length field */
	add	r2, r2, #4			/* add 4 for the line length offset (log2 16 bytes) */
	ldr	r4, =0x3ff
	ands	r4, r4, r1, lsr #3		/* r4 is the max number on the way size (right aligned) */
	clz	r5, r4				/* r5 is the bit position of the way size increment */
	ldr	r7, =0x7fff
	ands	r7, r7, r1, lsr #13		/* r7 is the max number of the index size (right aligned) */
.cloop2:
	mov	r9, r4				/* r9 working copy of the max way size (right aligned) */
.cloop3:
	orr	r11, r10, r9, lsl r5		/* factor in the way number and cache number into r11 */
	orr	r11, r11, r7, lsl r2		/* factor in the index number */

	mcr	p15, 0, r11, c7, c10, 2		// clean by set/way

	subs	r9, r9, #1			/* decrement the way number */
	bge	.cloop3
	subs	r7, r7, #1			/* decrement the index */
	bge	.cloop2
.cskip:
	add	r10, r10, #2			/* increment the cache number */
	cmp	r3, r10
	bgt	.cloop1

.cfinished:
	mov	r10, #0				/* swith back to cache level 0 */
	mcr	p15, 2, r10, c0, c0, 0		/* select current cache level in cssr */
	dsb
	isb

	pop {r0-r12, lr}
	bx	lr

/*
 *************************************************************************
 *
 * cleaninvalidate_dcache - clen & invalidate the entire d-cache by set/way
 *
 * Note: for Cortex-A53, there is no cp instruction for invalidating
 * the whole D-cache. Need to invalidate each line.
 *
  *
 *************************************************************************
 */

__cleaninvalidate_dcache:
	push	{r0 - r12, lr}

	mrc	p15, 1, r0, c0, c0, 1		/* read CLIDR */
	ands	r3, r0, #0x7000000
	mov	r3, r3, lsr #23			/* cache level value (naturally aligned) */
	beq	.cifinished
	mov	r10, #0				/* start with level 0 */
.ciloop1:
	add	r2, r10, r10, lsr #1		/* work out 3xcachelevel */
	mov	r1, r0, lsr r2			/* bottom 3 bits are the Cache type for this level */
	and	r1, r1, #7			/* get those 3 bits alone */
	cmp	r1, #2
	blt	.ciskip				/* no cache or only instruction cache at this level */
	mcr	p15, 2, r10, c0, c0, 0		/* write the Cache Size selection register */
	isb					/* isb to sync the change to the CacheSizeID reg */
	mrc	p15, 1, r1, c0, c0, 0		/* reads current Cache Size ID register */
	and	r2, r1, #7			/* extract the line length field */
	add	r2, r2, #4			/* add 4 for the line length offset (log2 16 bytes) */
	ldr	r4, =0x3ff
	ands	r4, r4, r1, lsr #3		/* r4 is the max number on the way size (right aligned) */
	clz	r5, r4				/* r5 is the bit position of the way size increment */
	ldr	r7, =0x7fff
	ands	r7, r7, r1, lsr #13		/* r7 is the max number of the index size (right aligned) */
.ciloop2:
	mov	r9, r4				/* r9 working copy of the max way size (right aligned) */
.ciloop3:
	orr	r11, r10, r9, lsl r5		/* factor in the way number and cache number into r11 */
	orr	r11, r11, r7, lsl r2		/* factor in the index number */

	mcr	p15, 0, r11, c7, c6, 2		// invalidate by set/way
	mcr	p15, 0, r11, c7, c10, 2		// clean by set/way
	//mcr	p15, 0, r11, c7, c12, 2		// clean and invilidate by set/way --> undefined instruction thrown! ??????

	subs	r9, r9, #1			/* decrement the way number */
	bge	.ciloop3
	subs	r7, r7, #1			/* decrement the index */
	bge	.ciloop2
.ciskip:
	add	r10, r10, #2			/* increment the cache number */
	cmp	r3, r10
	bgt	.ciloop1

.cifinished:
	mov	r10, #0				/* swith back to cache level 0 */
	mcr	p15, 2, r10, c0, c0, 0		/* select current cache level in cssr */
	dsb
	isb

	pop {r0-r12, lr}
	bx	lr


src/asm/cache32.s
/***********************************************************************************************************************
 * Raspberry Pi3 cache operation functions
 * cache opeartions are needed as the MMU need an active MMU and several
 * cross core/cross GPU operations need to clean/invalite the cache to se the
 * most actual data in the memory
 * TODO: actuall always the whole cache gets cleaned/invalidated, for better performance
 *       this should be enabled for specific address regions.
 *
 * Copyright (c) 2019 by the authors
 *
 * Author: André Borrmann
 * License: Apache License 2.0
 **********************************************************************************************************************/

.global __invalidate_dcache
.global __clean_dcache
.global __cleaninvalidate_dcache

/*
 *************************************************************************
 *
 * invalidate_dcache - invalidate the entire d-cache by set/way
 *
 * Note: for Cortex-A53, there is no cp instruction for invalidating
 * the whole D-cache. Need to invalidate each line.
 *
 *************************************************************************
 */

__invalidate_dcache:
	push	{r0 - r12, lr}

	mrc	p15, 1, r0, c0, c0, 1		/* read CLIDR */
	ands	r3, r0, #0x7000000
	mov	r3, r3, lsr #23			/* cache level value (naturally aligned) */
	beq	.ifinished
	mov	r10, #0				/* start with level 0 */
.iloop1:
	add	r2, r10, r10, lsr #1		/* work out 3xcachelevel */
	mov	r1, r0, lsr r2			/* bottom 3 bits are the Cache type for this level */
	and	r1, r1, #7			/* get those 3 bits alone */
	cmp	r1, #2
	blt	.iskip				/* no cache or only instruction cache at this level */
	mcr	p15, 2, r10, c0, c0, 0		/* write the Cache Size selection register */
	isb					/* isb to sync the change to the CacheSizeID reg */
	mrc	p15, 1, r1, c0, c0, 0		/* reads current Cache Size ID register */
	and	r2, r1, #7			/* extract the line length field */
	add	r2, r2, #4			/* add 4 for the line length offset (log2 16 bytes) */
	ldr	r4, =0x3ff
	ands	r4, r4, r1, lsr #3		/* r4 is the max number on the way size (right aligned) */
	clz	r5, r4				/* r5 is the bit position of the way size increment */
	ldr	r7, =0x7fff
	ands	r7, r7, r1, lsr #13		/* r7 is the max number of the index size (right aligned) */
.iloop2:
	mov	r9, r4				/* r9 working copy of the max way size (right aligned) */
.iloop3:
	orr	r11, r10, r9, lsl r5		/* factor in the way number and cache number into r11 */
	orr	r11, r11, r7, lsl r2		/* factor in the index number */

	mcr	p15, 0, r11, c7, c6, 2		// invalidate by set/way

	subs	r9, r9, #1			/* decrement the way number */
	bge	.iloop3
	subs	r7, r7, #1			/* decrement the index */
	bge	.iloop2
.iskip:
	add	r10, r10, #2			/* increment the cache number */
	cmp	r3, r10
	bgt	.iloop1

.ifinished:
	mov	r10, #0				/* swith back to cache level 0 */
	mcr	p15, 2, r10, c0, c0, 0		/* select current cache level in cssr */
	dsb
	isb

	dmb

	pop {r0-r12, pc}


/*
 *************************************************************************
 *
 * clean_dcache - clean the entire d-cache by set/way
 *
 * Note: for Cortex-A53, there is no cp instruction for invalidating
 * the whole D-cache. Need to invalidate each line.
 *
 *************************************************************************
 */

__clean_dcache:
	push	{r0 - r12, lr}

	mrc	p15, 1, r0, c0, c0, 1		/* read CLIDR */
	ands	r3, r0, #0x7000000
	mov	r3, r3, lsr #23			/* cache level value (naturally aligned) */
	beq	.cfinished
	mov	r10, #0				/* start with level 0 */
.cloop1:
	add	r2, r10, r10, lsr #1		/* work out 3xcachelevel */
	mov	r1, r0, lsr r2			/* bottom 3 bits are the Cache type for this level */
	and	r1, r1, #7			/* get those 3 bits alone */
	cmp	r1, #2
	blt	.cskip				/* no cache or only instruction cache at this level */
	mcr	p15, 2, r10, c0, c0, 0		/* write the Cache Size selection register */
	isb					/* isb to sync the change to the CacheSizeID reg */
	mrc	p15, 1, r1, c0, c0, 0		/* reads current Cache Size ID register */
	and	r2, r1, #7			/* extract the line length field */
	add	r2, r2, #4			/* add 4 for the line length offset (log2 16 bytes) */
	ldr	r4, =0x3ff
	ands	r4, r4, r1, lsr #3		/* r4 is the max number on the way size (right aligned) */
	clz	r5, r4				/* r5 is the bit position of the way size increment */
	ldr	r7, =0x7fff
	ands	r7, r7, r1, lsr #13		/* r7 is the max number of the index size (right aligned) */
.cloop2:
	mov	r9, r4				/* r9 working copy of the max way size (right aligned) */
.cloop3:
	orr	r11, r10, r9, lsl r5		/* factor in the way number and cache number into r11 */
	orr	r11, r11, r7, lsl r2		/* factor in the index number */

	mcr	p15, 0, r11, c7, c10, 2		// clean by set/way

	subs	r9, r9, #1			/* decrement the way number */
	bge	.cloop3
	subs	r7, r7, #1			/* decrement the index */
	bge	.cloop2
.cskip:
	add	r10, r10, #2			/* increment the cache number */
	cmp	r3, r10
	bgt	.cloop1

.cfinished:
	mov	r10, #0				/* swith back to cache level 0 */
	mcr	p15, 2, r10, c0, c0, 0		/* select current cache level in cssr */
	dsb
	isb

	dmb

	pop {r0-r12, pc}

/*
 *************************************************************************
 *
 * cleaninvalidate_dcache - clen & invalidate the entire d-cache by set/way
 *
 * Note: for Cortex-A53, there is no cp instruction for invalidating
 * the whole D-cache. Need to invalidate each line.
 *
  *
 *************************************************************************
 */

__cleaninvalidate_dcache:
	push	{r0 - r12, lr}

	mrc	p15, 1, r0, c0, c0, 1		/* read CLIDR */
	ands	r3, r0, #0x7000000
	mov	r3, r3, lsr #23			/* cache level value (naturally aligned) */
	beq	.cifinished
	mov	r10, #0				/* start with level 0 */
.ciloop1:
	add	r2, r10, r10, lsr #1		/* work out 3xcachelevel */
	mov	r1, r0, lsr r2			/* bottom 3 bits are the Cache type for this level */
	and	r1, r1, #7			/* get those 3 bits alone */
	cmp	r1, #2
	blt	.ciskip				/* no cache or only instruction cache at this level */
	mcr	p15, 2, r10, c0, c0, 0		/* write the Cache Size selection register */
	isb					/* isb to sync the change to the CacheSizeID reg */
	mrc	p15, 1, r1, c0, c0, 0		/* reads current Cache Size ID register */
	and	r2, r1, #7			/* extract the line length field */
	add	r2, r2, #4			/* add 4 for the line length offset (log2 16 bytes) */
	ldr	r4, =0x3ff
	ands	r4, r4, r1, lsr #3		/* r4 is the max number on the way size (right aligned) */
	clz	r5, r4				/* r5 is the bit position of the way size increment */
	ldr	r7, =0x7fff
	ands	r7, r7, r1, lsr #13		/* r7 is the max number of the index size (right aligned) */
.ciloop2:
	mov	r9, r4				/* r9 working copy of the max way size (right aligned) */
.ciloop3:
	orr	r11, r10, r9, lsl r5		/* factor in the way number and cache number into r11 */
	orr	r11, r11, r7, lsl r2		/* factor in the index number */

	mcr	p15, 0, r11, c7, c6, 2		// invalidate by set/way
	mcr	p15, 0, r11, c7, c10, 2		// clean by set/way
	//mcr	p15, 0, r11, c7, c12, 2		// clean and invilidate by set/way --> undefined instruction thrown! ??????

	subs	r9, r9, #1			/* decrement the way number */
	bge	.ciloop3
	subs	r7, r7, #1			/* decrement the index */
	bge	.ciloop2
.ciskip:
	add	r10, r10, #2			/* increment the cache number */
	cmp	r3, r10
	bgt	.ciloop1

.cifinished:
	mov	r10, #0				/* swith back to cache level 0 */
	mcr	p15, 2, r10, c0, c0, 0		/* select current cache level in cssr */
	dsb
	isb

	dmb
	
	pop {r0-r12, pc}


src/asm/cache64.s
/***********************************************************************************************************************
 * Raspberry Pi3 cache operation functions
 * cache opeartions are needed as the MMU need an active MMU and several
 * cross core/cross GPU operations need to clean/invalite the cache to se the
 * most actual data in the memory
 * TODO: actuall always the whole cache gets cleaned/invalidated, for better performance
 *       this should be enabled for specific address regions.
 *
 * Copyright (c) 2019 by the authors
 *
 * Author: André Borrmann
 * License: Apache License 2.0
 **********************************************************************************************************************/

.global __flush_icache_range
.global __flush_dcache_range
.global __invalidate_dcache
.global __clean_dcache
.global __cleaninvalidate_dcache

.macro save_state
	sub		sp, sp, #256 // make place at the stack to store all register values
	
	stp		x0, x1, [sp, #16 * 0]
	stp     x2, x3, [sp, #16 * 1]
	stp		x4, x5, [sp, #16 * 2]
	stp		x6, x7, [sp, #16 * 3]
	stp		x8, x9, [sp, #16 * 4]
	stp		x10, x11, [sp, #16 * 5]
	stp		x12, x13, [sp, #16 * 6]
	stp		x14, x15, [sp, #16 * 7]
	stp		x16, x17, [sp, #16 * 8]
	stp		x18, x19, [sp, #16 * 9]
	stp		x20, x21, [sp, #16 * 10]
	stp		x22, x23, [sp, #16 * 11]
	stp		x24, x25, [sp, #16 * 12]
	stp		x26, x27, [sp, #16 * 13]
	stp		x28, x29, [sp, #16 * 14]
	str     x30, [sp, #16 * 15]
.endm

.macro restore_state
	ldp		x0, x1, [sp, #16 * 0]
	ldp     x2, x3, [sp, #16 * 1]
	ldp		x4, x5, [sp, #16 * 2]
	ldp		x6, x7, [sp, #16 * 3]
	ldp		x8, x9, [sp, #16 * 4]
	ldp		x10, x11, [sp, #16 * 5]
	ldp		x12, x13, [sp, #16 * 6]
	ldp		x14, x15, [sp, #16 * 7]
	ldp		x16, x17, [sp, #16 * 8]
	ldp		x18, x19, [sp, #16 * 9]
	ldp		x20, x21, [sp, #16 * 10]
	ldp		x22, x23, [sp, #16 * 11]
	ldp		x24, x25, [sp, #16 * 12]
	ldp		x26, x27, [sp, #16 * 13]
	ldp		x28, x29, [sp, #16 * 14]
	ldr     x30, [sp, #16 * 15]

	add		sp, sp, #256 // free the stack as it is no longer needed
.endm


/**************************************************************************
 * read the minimum d-cache line size
 * Curruptile registers: x0, x1
 * returns x0 = cache line size
 **************************************************************************/
__get_dcache_line_size:
	mrs		x0, ctr_el0
	nop
	ubfm    x0, x0, #16, #19 // get bit's 16:19 and shift to the right (cache line encoding)
	mov     x1, #4           // bytes per word
	lsl     x0, x1, x0       // actual cache line size
	ret

/**************************************************************************
 * read the minimum i-cache line size
 * Corruptile registers: x0, x1
 * returns x0 = cache line size
 **************************************************************************/
__get_icache_line_size:
	mrs		x0, ctr_el0
	nop
	and    x0, x0, #0xF // get bit's 0:3 (cache line encoding)
	mov     x1, #4           // bytes per word
	lsl     x0, x1, x0       // actual cache line size
	ret

/**************************************************************************
 *
 * invalidate_dcache - invalidate the entire d-cache by set/way
 * 
 **************************************************************************/
__invalidate_dcache:
	save_state

	dsb	sy				// ensure ordering with previous memory accesses
	mrs	x0, clidr_el1			// read clidr
	and	x3, x0, #0x7000000		// extract loc from clidr
	lsr	x3, x3, #23			// left align loc bit field
	cbz	x3, ifinished			// if loc is 0, then no need to clean
	mov	x10, #0				// start clean at cache level 0
iloop1:
	add	x2, x10, x10, lsr #1		// work out 3x current cache level
	lsr	x1, x0, x2			// extract cache type bits from clidr
	and	x1, x1, #7			// mask of the bits for current cache only
	cmp	x1, #2				// see what cache we have at this level
	b.lt	iskip				// skip if no cache, or just i-cache
	msr	csselr_el1, x10			// select current cache level in csselr
	isb					// isb to sych the new cssr&csidr
	mrs	x1, ccsidr_el1			// read the new ccsidr
	and	x2, x1, #7			// extract the length of the cache lines
	add	x2, x2, #4			// add 4 (line length offset)
	mov	x4, #0x3ff
	and	x4, x4, x1, lsr #3		// find maximum number on the way size
	clz	w5, w4				// find bit position of way size increment
	mov	x7, #0x7fff
	and	x7, x7, x1, lsr #13		// extract max number of the index size
iloop2:
	mov	x9, x4				// create working copy of max way size
iloop3:
	lsl	x6, x9, x5
	orr	x11, x10, x6			// factor way and cache number into x11
	lsl	x6, x7, x2
	orr	x11, x11, x6			// factor index number into x11
	dc	isw, x11			// invalidate by set/way
	subs	x9, x9, #1			// decrement the way
	b.ge	iloop3
	subs	x7, x7, #1			// decrement the index
	b.ge	iloop2
iskip:
	add	x10, x10, #2			// increment cache number
	cmp	x3, x10
	b.gt	iloop1
ifinished:
	mov	x10, #0				// swith back to cache level 0
	msr	csselr_el1, x10			// select current cache level in csselr
	dsb	sy
	isb

	restore_state
	ret

/*
 *************************************************************************
 *
 * clean_dcache - clean the entire d-cache by set/way
 *
 * Note: for Cortex-A53, there is no cp instruction for invalidating
 * the whole D-cache. Need to invalidate each line.
 *
 *************************************************************************
 */

__clean_dcache:
	save_state

	dsb	sy				// ensure ordering with previous memory accesses
	mrs	x0, clidr_el1			// read clidr
	and	x3, x0, #0x7000000		// extract loc from clidr
	lsr	x3, x3, #23			// left align loc bit field
	cbz	x3, cfinished			// if loc is 0, then no need to clean
	mov	x10, #0				// start clean at cache level 0
cloop1:
	add	x2, x10, x10, lsr #1		// work out 3x current cache level
	lsr	x1, x0, x2			// extract cache type bits from clidr
	and	x1, x1, #7			// mask of the bits for current cache only
	cmp	x1, #2				// see what cache we have at this level
	b.lt	cskip				// skip if no cache, or just i-cache
	msr	csselr_el1, x10			// select current cache level in csselr
	isb					// isb to sych the new cssr&csidr
	mrs	x1, ccsidr_el1			// read the new ccsidr
	and	x2, x1, #7			// extract the length of the cache lines
	add	x2, x2, #4			// add 4 (line length offset)
	mov	x4, #0x3ff
	and	x4, x4, x1, lsr #3		// find maximum number on the way size
	clz	w5, w4				// find bit position of way size increment
	mov	x7, #0x7fff
	and	x7, x7, x1, lsr #13		// extract max number of the index size
cloop2:
	mov	x9, x4				// create working copy of max way size
cloop3:
	lsl	x6, x9, x5
	orr	x11, x10, x6			// factor way and cache number into x11
	lsl	x6, x7, x2
	orr	x11, x11, x6			// factor index number into x11
	dc	csw, x11			// clean by set/way
	subs	x9, x9, #1			// decrement the way
	b.ge	cloop3
	subs	x7, x7, #1			// decrement the index
	b.ge	cloop2
cskip:
	add	x10, x10, #2			// increment cache number
	cmp	x3, x10
	b.gt	cloop1
cfinished:
	mov	x10, #0				// swith back to cache level 0
	msr	csselr_el1, x10			// select current cache level in csselr
	dsb	sy
	isb

	restore_state
	ret

/**************************************************************************
 *
 * cleaninvalidate_dcache - clean & invalidate the entire d-cache by set/way
 *
 * Note: for Cortex-A53, there is no cp instruction for invalidating
 * the whole D-cache. Need to invalidate each line.
 *
 *
 **************************************************************************/

__cleaninvalidate_dcache:
	save_state

	dsb	sy				// ensure ordering with previous memory accesses
	mrs	x0, clidr_el1			// read clidr
	and	x3, x0, #0x7000000		// extract loc from clidr
	lsr	x3, x3, #23			// left align loc bit field
	cbz	x3, finished			// if loc is 0, then no need to clean
	mov	x10, #0				// start clean at cache level 0
loop1:
	add	x2, x10, x10, lsr #1		// work out 3x current cache level
	lsr	x1, x0, x2			// extract cache type bits from clidr
	and	x1, x1, #7			// mask of the bits for current cache only
	cmp	x1, #2				// see what cache we have at this level
	b.lt	skip				// skip if no cache, or just i-cache
	msr	csselr_el1, x10			// select current cache level in csselr
	isb					// isb to sych the new cssr&csidr
	mrs	x1, ccsidr_el1			// read the new ccsidr
	and	x2, x1, #7			// extract the length of the cache lines
	add	x2, x2, #4			// add 4 (line length offset)
	mov	x4, #0x3ff
	and	x4, x4, x1, lsr #3		// find maximum number on the way size
	clz	w5, w4				// find bit position of way size increment
	mov	x7, #0x7fff
	and	x7, x7, x1, lsr #13		// extract max number of the index size
loop2:
	mov	x9, x4				// create working copy of max way size
loop3:
	lsl	x6, x9, x5
	orr	x11, x10, x6			// factor way and cache number into x11
	lsl	x6, x7, x2
	orr	x11, x11, x6			// factor index number into x11
	dc	cisw, x11			// clean & invalidate by set/way
	subs	x9, x9, #1			// decrement the way
	b.ge	loop3
	subs	x7, x7, #1			// decrement the index
	b.ge	loop2
skip:
	add	x10, x10, #2			// increment cache number
	cmp	x3, x10
	b.gt	loop1
finished:
	mov	x10, #0				// swith back to cache level 0
	msr	csselr_el1, x10			// select current cache level in csselr
	dsb	sy
	isb

	restore_state
	ret
